torch/_tensor.py

```diff
@@ -103,7 +110,24 @@ def __reduce_ex__(self, proto):
         if has_torch_function_unary(self):
             return handle_torch_function(Tensor.__reduce_ex__, (self,), self, proto)
         func, args = self._reduce_ex_internal(proto)
-        return (_rebuild_from_type, (func, type(self), args, self.__dict__))
+        # Get the state of the python subclass
+        # This loosely mimicks the function on the object class but since Tensor do not inherit
+        # from it, we cannot call that function directly
+        # https://github.com/python/cpython/blob/c83919bd635f4433f1c6ae8504996a9fe3c215e5/Objects/typeobject.c#L4891
+        if hasattr(self, "__getstate__"):
+            state = self.__getstate__()
+        else:
+            # Contrary to the cpython function, we unify dict and slots into a single dictionary
+            # instead of a tuple of two dictionaries
+            if hasattr(self.__class__, "__slotnames__"):
+                attr_to_save = self.__class__.__slotnames__
+            else:
+                import copyreg
+                attr_to_save = copyreg._slotnames(self.__class__)
+            # Tensors always have a __dict__
+            attr_to_save += ["__dict__"]
+            state = {name: getattr(self, name) for name in attr_to_save if hasattr(self, name)}
```

Comment:
Is getattr/hasattr here correct semantics in the presence of `__getattr__` and friends?

Reply:
For the getattr:
- For the `__dict__` cpython gets it as one element [here](https://github.com/python/cpython/blob/c83919bd635f4433f1c6ae8504996a9fe3c215e5/Objects/typeobject.c#L4912) and returns it as-is [here](https://github.com/python/cpython/blob/c83919bd635f4433f1c6ae8504996a9fe3c215e5/Objects/typeobject.c#L4919). We do the same here using getattr to get the full dict in one go.
- For each slot element, cpython calls getattr() for each of them, skipping if it doesn't exit [here](https://github.com/python/cpython/blob/c83919bd635f4433f1c6ae8504996a9fe3c215e5/Objects/typeobject.c#L4969) just like we do.

For the setting:
- In cpython, all elements in the `__dict__` are set [here](https://github.com/python/cpython/blob/9ad8f109ac037ac088e09dcd2da523322c5caf34/Modules/_pickle.c#L6710) one after the other. In the code above, I only set `ret.__dict__` directly because that's what the old code was doing. Should I update it to call `setattr()` for each element in the dict?
- For other slot elements, cpython does a setattr() for each [here](https://github.com/python/cpython/blob/9ad8f109ac037ac088e09dcd2da523322c5caf34/Modules/_pickle.c#L6732) just like we do.

Reply:
I guess my question was more, getattr respects `__getattr__`

```
>>> class A:
...   def __getattr__(self, x):
...     print(x)
... 
>>> A().x
x
>>> getattr(A(), 'x')
x
```

and I wasn't sure if `_PyObject_LookupAttr` also respects the magic method or not.

Reply:
It does respect both `tp_getattr` and `tp_getattro` which are the ones delegating to `__getattr__` or `__getattribute__ `.